### PR TITLE
#837 Devcontainer cli add label

### PR DESCRIPTION
**Pull request by prathameshzarkar9**
[Issue #832](https://github.com/devcontainers/cli/issues/837)
Adding labels to containers build by devcontainer cli

### CODE REVIEW

File: src/spec-node/containerFeatures.ts [hunk 1]
```diff
@@ -113,6 +113,7 @@ export async function extendImage(params: DockerResolverParameters, config: Subs
 		'--target', featureBuildInfo.overrideTarget,
 		'-f', dockerfilePath,
 		...additionalImageNames.length > 0 ? additionalImageNames.map(name => ['-t', name]).flat() : ['-t', updatedImageName],
+		...imageLabels.length > 0 ? imageLabels.map(label => ['--label', label]).flat() : ['--label', imageLabels.toString()],
```
Comment (samruddhikhandale) [inline src/spec-node/containerFeatures.ts:116]: If there are no labels, why don we want to generates a single `--label` argument with an empty value?

File: src/spec-node/devContainers.ts [hunk 2]
```diff
@@ -146,6 +147,7 @@ export async function createDockerParams(options: ProvisionOptions, disposables:
 		secretsP,
 		buildxPlatform: options.buildxPlatform,
 		buildxPush: options.buildxPush,
+		buildxLabel: options.buildxLabel,
```
Comment (samruddhikhandale) [inline src/spec-node/devContainers.ts:150]: I am also seeing many failures, can you fix those? 👇 

image

File: src/spec-node/devContainersSpecCLI.ts [hunk 3]
```diff
@@ -508,6 +510,7 @@ function buildOptions(y: Argv) {
 		'buildkit': { choices: ['auto' as 'auto', 'never' as 'never'], default: 'auto' as 'auto', description: 'Control whether BuildKit should be used' },
 		'platform': { type: 'string', description: 'Set target platforms.' },
 		'push': { type: 'boolean', default: false, description: 'Push to a container registry.' },
+		'label': { type: 'string', description: 'Provide key and value for the metadata of the container. (<key1>=<value1> <key2>=<value2> )' },
```
Comment (samruddhikhandale) [inline src/spec-node/devContainersSpecCLI.ts:513]:
```suggestion
		'label': { type: 'string', description:  Provide key and value configuration that adds metadata to an image.' },
```

File: src/spec-node/singleContainer.ts [hunk 4]
```diff
@@ -119,15 +119,16 @@ async function setupContainer(container: ContainerDetails, params: DockerResolve
...
+export async function buildNamedImageAndExtend(params: DockerResolverParameters, configWithRaw: SubstitutedConfig<DevContainerFromDockerfileConfig | DevContainerFromImageConfig>, additionalFeatures: Record<string, string | boolean | Record<string, string | boolean>>, canAddLabelsToContainer: boolean, argImageNames?: string[], imageLabels?: string[]): Promise<{ updatedImageName: string[]; imageMetadata: SubstitutedConfig<ImageMetadataEntry[]>; imageDetails: () => Promise<ImageDetails>; labels?: Record<string, string> }> {
 	const { config } = configWithRaw;
 	const imageNames = argImageNames ?? [getDefaultName(config, params)];
+	const imagelabel = imageLabels ?? [getDefaultName(config, params)];
```
Comment (samruddhikhandale) [inline src/spec-node/singleContainer.ts:125]:
```suggestion
	const imagelabel = imageLabels;
```

Why do we want `[getDefaultName(config, params)]` as default if image labels doesn't exist?

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/spec-node/devContainersSpecCLI.ts (hunk 3 above)
Comment (prathameshzarkar9) [reply src/spec-node/devContainersSpecCLI.ts:513]: done

File: src/spec-node/singleContainer.ts (hunk 4 above)
Comment (prathameshzarkar9) [reply src/spec-node/singleContainer.ts:125]: done

File: src/spec-node/devContainers.ts (hunk 2 above)
Comment (prathameshzarkar9) [reply src/spec-node/devContainers.ts:150]: fixed all the test cases

File: src/spec-node/containerFeatures.ts [hunk 5]
```diff
@@ -28,7 +28,7 @@ export const getSafeId = (str: string) => str
 	.replace(/^[\d_]+/g, '_')
 	.toUpperCase();
```
[the remark this answers is not in the document]
Comment (prathameshzarkar9) [reply src/spec-node/containerFeatures.ts:30]: fixed

File: src/spec-node/containerFeatures.ts (hunk 1 above)
Comment (prathameshzarkar9) [reply src/spec-node/containerFeatures.ts]: changed to empty if --label is not there

**Comment by prathameshzarkar9** [comment]
hi @samruddhikhandale ,
All the test cases are not succeed. Kindly review the PR once.

File: src/spec-node/utils.ts [hunk 6]
```diff
@@ -116,6 +116,7 @@ export interface DockerResolverParameters {
 	experimentalFrozenLockfile?: boolean;
 	buildxPlatform: string | undefined;
 	buildxPush: boolean;
+	buildxLabel: string | undefined;
```
Comment (chrmarti) [inline src/spec-node/utils.ts:119]: Labels don't depend on `buildx` being available. I suggest to make this similar to `additionalMounts: Mount[];` above, e.g.: `additionalLabels: string[];`.

File: src/spec-node/singleContainer.ts [hunk 7]
```diff
@@ -119,15 +119,16 @@ async function setupContainer(container: ContainerDetails, params: DockerResolve
...
 	return 'image' in config && config.image ? config.image : getFolderImageName(params.common);
 }
-export async function buildNamedImageAndExtend(params: DockerResolverParameters, configWithRaw: SubstitutedConfig<DevContainerFromDockerfileConfig | DevContainerFromImageConfig>, additionalFeatures: Record<string, string | boolean | Record<string, string | boolean>>, canAddLabelsToContainer: boolean, argImageNames?: string[]): Promise<{ updatedImageName: string[]; imageMetadata: SubstitutedConfig<ImageMetadataEntry[]>; imageDetails: () => Promise<ImageDetails>; labels?: Record<string, string> }> {
+export async function buildNamedImageAndExtend(params: DockerResolverParameters, configWithRaw: SubstitutedConfig<DevContainerFromDockerfileConfig | DevContainerFromImageConfig>, additionalFeatures: Record<string, string | boolean | Record<string, string | boolean>>, canAddLabelsToContainer: boolean, argImageNames?: string[], imageLabels?: string[]): Promise<{ updatedImageName: string[]; imageMetadata: SubstitutedConfig<ImageMetadataEntry[]>; imageDetails: () => Promise<ImageDetails>; labels?: Record<string, string> }> {
```
Comment (chrmarti) [inline src/spec-node/singleContainer.ts:122]: The labels are part of the `params` object, the additional parameter might not be needed here.

File: src/spec-node/containerFeatures.ts [hunk 8]
```diff
@@ -28,7 +28,7 @@ export const getSafeId = (str: string) => str
...
 
-export async function extendImage(params: DockerResolverParameters, config: SubstitutedConfig<DevContainerConfig>, imageName: string, additionalImageNames: string[], additionalFeatures: Record<string, string | boolean | Record<string, string | boolean>>, canAddLabelsToContainer: boolean) {
+export async function extendImage(params: DockerResolverParameters, config: SubstitutedConfig<DevContainerConfig>, imageName: string, additionalImageNames: string[], imageLabels: string[], additionalFeatures: Record<string, string | boolean | Record<string, string | boolean>>, canAddLabelsToContainer: boolean) {
```
Comment (chrmarti) [inline src/spec-node/containerFeatures.ts:31]: The labels are part of the `params` object, the additional parameter might not be needed here.

**Review by chrmarti — CHANGES_REQUESTED** [review]
Thanks for the PR, left a few comments.

File: src/spec-node/utils.ts (hunk 6 above)
Comment (prathameshzarkar9) [reply src/spec-node/utils.ts:119]: Done

**Review by chrmarti — APPROVED** [review]
LGTM!

**Review by samruddhikhandale — APPROVED** [review]
Wonderful, thanks!

**Pull request merged**